app/src/common/shared/com/igalia/wolvic/ui/adapters/WebApp.java

```diff
@@ -0,0 +1,163 @@
+        mShortName = manifest.optString("short_name");
+        mScope = manifest.optString("scope");
+
+        int themeColor = Color.TRANSPARENT;
+        try {
+            String themeColorString = manifest.optString("theme_color");
+            themeColor = Color.parseColor(themeColorString);
+        } catch (IllegalArgumentException e) {
+            Log.w(LOGTAG, "Invalid theme_color value");
+        }
+        mThemeColor = themeColor;
+
+        int backgroundColor = Color.TRANSPARENT;
+        try {
+            String backgroundColorString = manifest.optString("background_color");
+            backgroundColor = Color.parseColor(backgroundColorString);
+        } catch (IllegalArgumentException e) {
+            Log.w(LOGTAG, "Invalid background_color value");
+        }
+        mBackgroundColor = backgroundColor;
+
+        // Algorithm for calculating Identity at https://www.w3.org/TR/appmanifest/#id-member
+        if (!StringUtils.isEmpty(mStartUrl)) {
+            URL startUrl = new URL(mStartUrl);
+            String id = manifest.optString("id");
+            URL identityUrl = new URL(startUrl.getProtocol(), startUrl.getHost(), startUrl.getPort(), id);
+            mIdentity = identityUrl.toString();
+        } else {
+            // Since Identity is used to uniquely identify the Web application,
+            // we treat its absence as an error.
+            throw new IOException("Unable to assign an identity to the Web App manifest");
+        }
+    }
+
+    @NonNull
+    public String getId() {
+        return mIdentity;
+    }
+
+    public String getName() {
+        return mName;
+    }
+
+    public String getShortName() {
+        return mShortName;
+    }
+
+    public String getScope() {
+        return mScope;
+    }
+
+    public String getStartUrl() {
+        return mStartUrl;
+    }
+
+    public int getThemeColor() {
+        return mThemeColor;
+    }
+
+    public int getBackgroundColor() {
+        return mBackgroundColor;
+    }
+
+    public void copyFrom(WebApp webApp) {
+        mIdentity = webApp.mIdentity;
+        mName = webApp.mName;
+        mShortName = webApp.mShortName;
+        mScope = webApp.mScope;
+        mStartUrl = webApp.mStartUrl;
+        mThemeColor = webApp.mThemeColor;
+        mBackgroundColor = webApp.mBackgroundColor;
+
+        mHashCode = OptionalInt.empty();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o)
+            return true;
+        if (o == null || getClass() != o.getClass())
+            return false;
+        WebApp webApp = (WebApp) o;
+        return mIdentity.equals(webApp.mIdentity) &&
+                mName.equals(webApp.mName) &&
+                mShortName.equals(webApp.mShortName) &&
+                mScope.equals(webApp.mScope) &&
+                mStartUrl.equals(webApp.mStartUrl) &&
+                mThemeColor == webApp.mThemeColor &&
+                mBackgroundColor == webApp.mBackgroundColor;
```

Comment:
Out of curiosity, I guess there could not be (in practice) two instances of WebApps with equal identities but differences in other fields. I was wondering whether it'd be enough to simply check for the id, but shrug. What you do is clearly safer anyway.

Reply:
In that case, the most recent manifest should take precedence.

I'm trying to follow this spec: https://www.w3.org/TR/appmanifest/#id-member

> The [identity](https://www.w3.org/TR/appmanifest/#dfn-identity) is used by user agents to uniquely identify the application universally. 
>
> When the user agent sees a manifest with an [identity](https://www.w3.org/TR/appmanifest/#dfn-identity) that does not correspond to an already-installed application, it SHOULD treat that manifest as a description of a distinct application, even if it is served from the same URL as that of another application.
>
>When the user agent sees a manifest where manifest["id"] is [equal](https://url.spec.whatwg.org/#concept-url-equals) with [exclude fragment true](https://url.spec.whatwg.org/#url-serializer-exclude-fragment) to the [identity](https://www.w3.org/TR/appmanifest/#dfn-identity) of an already-installed application, it SHOULD be used as a signal that this manifest is a replacement for the already-installed application's manifest, and not a distinct application, even if it is served from a different URL than the one seen previously.